emulator: Complex Types: Implement AddToCell mutation

marete opened the pull request:
This implements the AddToCell mutation.

- Unit tested and tests passing
- Supports atomic rollback
- Support Sum, Min and Max aggregations for non-negative Int64 numbers
- It is now possible to create an aggregate column family (supporting AddToCell or MergeToCell)

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/3)
<!-- Reviewable:end -->

marete committed:
emulator: expose find and erase map methods for the column family types.

marete committed:
emulator: table.h: Add some header files.

marete committed:
emulator: Implement a SetCell that records an undo log.

marete committed:
Complete implementation of Undo() for SetCell mutations.

marete committed:
emulator: Run clang-format on the subset of the files we have been changing.

marete committed:
emulator: Add a static method to Table that returns a default-constructed table.

marete committed:
emulator: testing: Create a table in the correct way.

Populate a schema protobuf with the minimal fields and call
Table::Create();

marete committed:
emulator: SetCell mutation: Various logic fixes for undo log.

marete committed:
emulator: rollback tests: fix test for basic functionality.

Thanks to Marek for the hint on the major part of the fix, which
involved copying a reference instead of a value!

marete committed:
emulator: Simplify some of transactions SetCell logic.

marete committed:
emulator: fix bug: Don't modify a copy of ColumnRow.

marete committed:
emulator: rollback tests: Implement additional testing helper functions.

- To check if a row exists
- To check if a column exists

marete committed:
emualtor: transaction rollback: Implement working test for RestoreValue.

This also involves a fix to the code due to a bug revealed by the
test!

marete committed:
emulator: rollback_tests.cc: run clang-format.

marete committed:
emulator: transaction rollback: Avoid copying the possibly very large cell values.

marete committed:
emulator: transaction rollback: SetCell: Implement working test for DeleteValue.

marete committed:
emulator: transaction rollback: Implement working DeleteRow test.

marete committed:
emulator: transaction rollback: Implement a DeleteFromFamily that updates the undo log.

Also, add code coverage output artefacts to ignored files.

marete committed:
emulator: transaction rollback: Implement undo for row deletions (DeleteFromFamily).

And be careful not to potentially very large values (use std::move on
the column qualifier and cell values.

marete committed:
emulator: RestoreRow undo log: Rename struct members more conventionally.

marete committed:
emulator: transaction rollback: Test basic functionality of a DeleteFromFamily which uses RowTransaction.

marete committed:
emulator: test rollback of a row deletion due to DeleteFromFamily.

And fix a code bug discovered by this test.

marete committed:
emulator: Fixes for review: Improve name of return varible.

In particular, make it clear that we are only examining the status,
and will not use the possibly valid status value (returned object).

marete committed:
emulator: Rename RestoreRow -> RestoreColumnFamilyRow.

marete committed:
emulator: style: Remove trailing underscore from struct member names.

marete committed:
emulator: Rename several iterator variables more appropriately.

marete committed:
emulator: Use std::move() to avoid a copy.

marete committed:
emulator: Rename another iterator properly.

marete committed:
emulator: Use std::move() to avoid another copy.

marete committed:
emulator: check and return from exceptional conditions first.

This is to remain in line with the rest of the code.

marete committed:
emualtor: Fix the naming of another iterator variable.

marete committed:
emulator: remove use of C++17 syntax feature (if with initializer).

We need to have the code compilable by a C++14 compiler.

marete committed:
emulator: rename another iterator variable more appropriately.

marete committed:
emulator: Add a comment justifying storing the column family iterator in the undo log.

In general, it is unsafe to store other kind of map iterators in the
undo log, since a subsequent mutation on the same row could delete the
underlying map, and invalidate the iterator. However, deleting a
column family is a schema operation that cannot happen while we hold
the table lock, therefore for column families only, it is safe to
store the iterators in the undo log.

marete committed:
emulator: transaction rollback: RestoreValue: Eliminate the storage of iterators in undo log.

marete committed:
emulator: actually remove iterator field from RestoreValue.

marete committed:
emulator: Correct and radically simplify SetCell undo code logic.

- Don't store iterators in the DeleteValue undo type (they may be
  invalidated by a subsequent mutation).

- Maintain the invariants that there is never an empty column or empty
  row by using the ColumnFamily::DeleteColumn() function to delete
  ranges of cells during undo. The latter takes care of deleting empty
  columns and empty rows, already.

- And therefore we can get rid of the DeleteRow and DeleteColumn undo
  types and associated code.

All tests continue to pass.

marete committed:
emulator: remove undo types that are no longer needed.

marete committed:
emulator: Modify the Delete* methods to return vector of deleted cells.

And also use std::move() to avoid copying potentially large cell values.

These change is needed to simplify the logic of preparing undo logs
for the Delete* mutations. In particular, we don't have to re-iterate
the table.

marete committed:
emualator: Implement DeleteFromColumn mutation using RowTransaction.

marete committed:
emulator: Test transaction DeleteFromColumn for basic functionality.

marete committed:
emulator: DeleteFromColumn basic functionality test: test that the column is actually deleted.

marete committed:
emulator: make it possible to test a vector of DeleteFromColumn row mutations.

This is needed to test DeleteFromColumn rollback.

marete committed:
emulator: Test transaction rollback for DeleteFromColumn.

marete committed:
emulator: TEST(TransactonRollback, DeleteFromColumnRollback): also check that the cells are properly restored.

marete committed:
emulator: Fix segmentation fault caused by invalidated iterator in DeleteRow.

Also implement DeleteFromFamily in terms of the new DeleteRow which
returns RestoreValue vectors (and therefore eliminiate the use of one
more undo log type).

marete committed:
emulator: clang-format.

marete committed:
emulator: transaction rollback: get rid of another undo log type.

This leaves us with just 2 undo log types: RestoreValue and
DeleteValue, a situation that is much simpler than before.

marete committed:
emulator: Implement a DeleteFromRow that supports rollback.

marete committed:
emulator: Test DeleteFromRow with rollback support and use it in MutateRow code.

marete committed:
emualtor: Remove now-unused undo type.

marete committed:
emulator: Complex Types: Implement AddToCell mutation.

This implements the AddToCell mutation.

* Unit tested and tests passing
* Supports atomic rollback
* Support Sum, Min and Max aggregations for non-negative Int64 numbers
* It is now possible to create an aggregate column family (supporting AddToCell or MergeToCell)

marete committed:
emulator: Fixes to avoid copying values that are possibly very large (cell contents).

As well as other fixes for review.

marete committed:
emulator: mutation transactions: use std::chrono:duration_cast.

Makes it easier and safer to change the resolution at a later date.

marete committed:
emulator: use invariant that rows are never empty to eliminate some code.

Preceding code erases all columns. The last column erase in
DeleteColumn will cause the row to be erased. So we don't have to
search for it again in the map and attempt to erase it.

marete committed:
emulator: optimization: make it possible to delete a column via a row iterator.

In this and other situations, this saves us a duplicate call to find
the row by saving and passing an iterator.

marete committed:
emulator: eliminate redundant storage of row_key in undo log.

The entire undo is in the context of a row, therefore all the methods
have access to a row key.

This also reduces the size of the undo log.

marete committed:
emulator: Use google-standard status builder everywhere in main code.

marete committed:
emulator: eliminate a C++-17ism.

marete committed:
emulator: remove non-Google-standard Status builders.

marete committed:
emulator: make some std::chrono usage more type-safe.

marete committed:
emulator: eliminate some code duplication.

marete committed:
emulator: SetCell: return overwritten value (if any).

This prepares the way to simplify the code that prepares the undo log,
and in particular it obviates the need to check to see if the
key (timestamp) existed before the SetCell mutation.

marete committed:
emulator: radically simplify undo log preparation code.

This is thanks to simply returning the old value (if any) rather than
performing almost endless iterations in this function! :)

marete commented:
To review this, having rebased it over emulator (the large PR) simply choose "Changes from 1 commit" which is the top commit as follows:

![image](https://github.com/user-attachments/assets/11e55861-8994-43ea-af86-8dbf1e5eb978)

PTAL. Thank you.

dopiera commented on google/cloud/bigtable/emulator/column_family.cc:
```diff
@@ -14,14 +14,55 @@
 namespace bigtable {
 namespace emulator {
 
+uint64_t BigEndianToUint64(std::string const& s) {
```
Why not use the implementation from `google/cloud/internal/big_endian.h`?

marete replied:
Done.
<!-- Reviewable comment -OQFGFwq9FcPRkLk84Qa:b-896fix -->
<!-- Sent from Reviewable.io -->

marete committed:
emulator: SetCell: set timestamp to server time if passed 0 or negative timestamp.

marete committed:
emulator: transaction rollback: Check that a SetCell with timestamp 0 can be correctly rolled back.

marete committed:
emulator: Implement ColumnFamily::DeleteTimeStamp().

This allows us to delete cells by specifying a specific timestamp
rather than a timestamp range which might otherwise involve possibly
non-type-safe calculations to determine the upper bound of the
timestamps (in particular doing the latter needs one to assume a
specific resolution for bigtable's timestamp storage).

marete committed:
emulator: transaction undo: radically simplify the code for DeleteValue.

marete committed:
emulator: rollback tests: perform some trivial refactoring.

dopiera commented on google/cloud/bigtable/emulator/column_family.h:
```diff
@@ -288,8 +295,53 @@ class ColumnFamily {
+    return new_value;
+  };
+
+  static std::string Sum_UpdateCell_BE_Uint64(std::string const& existing_value,
```
Please make sure that all names follow the [style guide](https://google.github.io/styleguide/cppguide.html#Function_Names)

marete replied:
Done.
<!-- Reviewable comment -OQEN0AZ8un9K0Ki3a5L:b-896fix -->
<!-- Sent from Reviewable.io -->

dopiera commented on google/cloud/bigtable/emulator/column_family.cc:
```diff
@@ -110,6 +151,30 @@ absl::optional<Cell> ColumnFamilyRow::DeleteTimeStamp(
+      auto column_row_it = column_it->second.find(timestamp);
+      if (column_row_it != column_it->second.end()) {
+        // We are updating an existing cell
+        update_value = UpdateCell_(column_row_it->second, value);
```
You're essentially rewriting the logic from `ColumnFamilyRow::SetCell()` and `ColumnFamily::SetCell()` here. IMO this is not good for two reasons:

1. we're performing the lookups in the maps twice
2. we're duplicating code and hence increase the risk of introducing different behaviors (think 0-timestamp handling for example)

IMO, rather than doing that, we should extend `ColumnFamilyRow::SetCell()` and `ColumnFamily::SetCell()` signatures with the update function.

marete replied:
Done.
<!-- Reviewable comment -OQiQwlE0JpvaFcyNn5F:b-896fix -->
<!-- Sent from Reviewable.io -->

dopiera commented on google/cloud/bigtable/emulator/table.cc:
```diff
@@ -395,9 +409,94 @@ bool Table::IsDeleteProtectedNoLock() const {
+  auto cf_value_type = cf.GetValueType();
+  if (!cf_value_type.has_value() ||
+      !cf_value_type.value().has_aggregate_type()) {
+    return Status(
```
Please use `GCP_ERROR_INFO().WithMetadata` and quote the value types of the mutation and CF.

marete replied:
Done.
<!-- Reviewable comment -OQJiF9S1Let4c0H7wze:b-896fix -->
<!-- Sent from Reviewable.io -->

dopiera commented on google/cloud/bigtable/emulator/table.cc:
```diff
@@ -75,9 +79,18 @@ Status Table::Construct(google::bigtable::admin::v2::Table schema) {
+    absl::optional<google::bigtable::admin::v2::Type> opt_value_type =
+        absl::nullopt;
+
+    // Support for complex types (Add_To_Cell aggregations, e.t.c.).
```
Wouldn't it be simpler if:

1. Instead of the ctor accepting an option you had a ctor accepting the type
2. You used either the default ctor or the ctor for the type
3. Rewrote the code to
```
if (column_family_def.second.has_value_type()) {
  column_families_.emplace(column_family_def.first,
                             std::make_shared<ColumnFamily>(column_family_def.second.value_type()));	
} else {
  column_families_.emplace(column_family_def.first,
                             std::make_shared<ColumnFamily>());	
}
```

marete replied:
This is indeed what I have done :). Done.
<!-- Reviewable comment -OQJhkwdCAUa02TNxnnC:b-8an8y4 -->
<!-- Sent from Reviewable.io -->

dopiera commented on google/cloud/bigtable/emulator/column_family.cc:
```diff
@@ -294,6 +359,38 @@ bool FilteredColumnFamilyStream::PointToFirstCellAfterRowChange() const {
+        UpdateCell_ = Max_UpdateCell_BE_Uint64;
+        break;
+      default:
+        break;
```
Silent failure is not great. We should also not crash the emulator on bad user data. I think we should do one of these:
1. assert (and make sure that graceful error reporting in case of unsupported types is present in the calling function)
2. convert this ctor to a `Create` member function returning a `StatusOr`

dopiera replied:
I strongly vouch for the latter so that we don't need to keep two pieces of code in sync.

marete replied:
Done.
<!-- Reviewable comment -OQJhf3y7hS5cM75K7gu:b-896fix -->
<!-- Sent from Reviewable.io -->

dopiera commented on google/cloud/bigtable/emulator/table.cc:
```diff
@@ -395,9 +409,94 @@ bool Table::IsDeleteProtectedNoLock() const {
+    default:
+      return Status(
+          StatusCode::kUnimplemented,
+          absl::StrFormat(
```
As mentioned in #2 please use `GCP_ERROR_INFO().WithMetadata`

marete replied:
Done.
<!-- Reviewable comment -OQJiaJ81E86IOrkqQXT:b-896fix -->
<!-- Sent from Reviewable.io -->

dopiera commented on google/cloud/bigtable/emulator/table.cc:
```diff
@@ -395,9 +409,94 @@ bool Table::IsDeleteProtectedNoLock() const {
+    }
+  }
+
+  if (!cell_existed) {
```
You will probably need to rebase this on top of your recent changes to follow the simplified code organization where you don't need to do this.

marete replied:
Done.
<!-- Reviewable comment -OQPl0Pn9Krx73b02R8o:b-896fix -->
<!-- Sent from Reviewable.io -->

dopiera commented on google/cloud/bigtable/emulator/column_family.cc:
```diff
@@ -294,6 +359,38 @@ bool FilteredColumnFamilyStream::PointToFirstCellAfterRowChange() const {
+    return;
+  }
+
+  // FIXME: We currently only support big-endian uint64
```
I initially got confused as well, but I think I finally understand what the story with "ordered" is. Here is my understanding:
Bigtable only supports signed, 64-bit integers encoded as twos complement. It supports both negative and positive numbers.

The comment from the protobuf refers to their ordering when they're serialized. I believe the authors meant to say that one can expect that comparing two strings with serialized numbers will yield the same comparison result as comparing the numbers only for non-negative numbers. This is only an indication to the users that they should not compare the raw strings if they represent negative numbers. Irrespective of that, you can store negative numbers in Bigtable.

Hence, please use `std::int64_t`. Also, please make sure you use `std::int64_t` rather than `int64_t` throughout your change for consistency with the rest of the code.

marete replied:
Done.
<!-- Reviewable comment -OQIsiPB3T6fGMPlY9GK:b-896fix -->
<!-- Sent from Reviewable.io -->

dopiera commented on google/cloud/bigtable/emulator/table.cc:
```diff
@@ -395,9 +409,94 @@ bool Table::IsDeleteProtectedNoLock() const {
-  return UnimplementedError(
-      "Unsupported mutation type.",
-      GCP_ERROR_INFO().WithMetadata("mutation", add_to_cell.DebugString()));
+  auto status = table_->FindColumnFamily(add_to_cell);
```
I think we should have a separation of duties between `RowTransaction` and other classes. To me, it is:
* `RowTransaction` only takes care of creating the undo log
* all other classes (`Table`, `ColumnFamily`, `ColumnFamilyRow`, etc.) handle the updating of the data
Can you please make sure that we follow this guideline?

marete replied:
Done.
<!-- Reviewable comment -OQiUeWQ8qzbG7EW9GYp:b-896fix -->
<!-- Sent from Reviewable.io -->

dopiera commented on google/cloud/bigtable/emulator/table.cc:
```diff
@@ -395,9 +409,94 @@ bool Table::IsDeleteProtectedNoLock() const {
+  switch (add_to_cell.input().kind_case()) {
+    case google::bigtable::v2::Value::kIntValue:
+      if (!add_to_cell.input().has_int_value() ||
+          add_to_cell.input().int_value() < 0) {
```
As outlined elsewhere in the review, I believe this is excessive

marete replied:
Done.
<!-- Reviewable comment -OQJicB52bPFyOHJ6ony:b-896fix -->
<!-- Sent from Reviewable.io -->

dopiera commented on google/cloud/bigtable/emulator/table.cc:
```diff
@@ -75,9 +79,18 @@ Status Table::Construct(google::bigtable::admin::v2::Table schema) {
         GCP_ERROR_INFO().WithMetadata("schema", schema_.DebugString()));
   }
   for (auto const& column_family_def : schema_.column_families()) {
+    absl::optional<google::bigtable::admin::v2::Type> opt_value_type =
```
We should also hook into `ModifyColumnFamilies`

marete replied:
Done.
<!-- Reviewable comment -OQKd8M751dveKHg0caU:b-896fix -->
<!-- Sent from Reviewable.io -->

marete committed:
emulator: fixes for clang-tidy and clang-format.

Restricted only to changes I have made to this PR.

Tested (all tests pass).

marete committed:
emulator: AddToCell: Use CamelCase for function names.

marete committed:
emulator: ColumnFamily, AddToCell: Various important fixes from review.

- Accept any std::int64_t input
- Use already available functions to encode the integers into
  big-endian format.
- Use the Google-Standard functions to report error Statuses.
- ColumnFamily: Replace defective non-default constructor of aggregate
- column families with a function that can return an error status.
- Mechanical formatting fixes.

marete committed:
emulator: ModifyColumnFamilies: Add support for the creation of aggregate column families.

marete committed:
emulator: ModifyColumnFamilies: Disallow the update of value_type.

value_type (the kind of data stored in the column family and its
encoding) can still be set in during column family creation.

marete committed:
emulator: rollback_tests.cc: Fix case of function names,

Use CamelCase as the Google style guide requires.

marete committed:
emulator: column_family.h: Fix mis-merge from rebase operation.

marete committed:
emulator: Fix ColumnFamily::DeleteRow: Broken by mis-merge during rebase.

marete committed:
emulator: Fix typo.

marete committed:
emulator: mechanical formatting fixes by ./ci/cloudbuild/build.sh -t checkers-ci.

marete committed:
emulator: SetCell: Eliminate duplicate map key lookup for complex mutations.

marete committed:
emulator: AddToCell: Update to use newer more correct infrastructure.

- Use new SetCell which takes an update function (to be used in the
  case of aggregate/complex types).
- Fix handling of unset fields which are critical to the
  mutation (timestamp, column qualifier).
- Fix setup and preparation of the undo log to match the modern way we
  do it in SetCell.

marete committed:
emulator: get rid of a lint warning.

marete reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/3)* status: 0 of 11 files reviewed, 12 unresolved discussions (waiting on @dopiera)



<!-- Sent from Reviewable.io -->

marete reviewed (commented):
PTAL.

> *[Reviewable](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/3#-:-OQiUp-3BWf1FpJVHEKW:b-83330m)* status: 0 of 11 files reviewed, 12 unresolved discussions (waiting on @dopiera)



<!-- Sent from Reviewable.io -->

marete committed:
emulator: mechanical fixes by ci tool.

dopiera reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/3)* status: 0 of 11 files reviewed, 14 unresolved discussions

___
*[`google/cloud/bigtable/emulator/table.cc` line 194 at r11](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/3#-OQs3YGlBtcbx04fRmwl:-OQs3YGlBtcbx04fRmwm:bvsh4l2) ([raw file](https://github.com/Unoperate/google-cloud-cpp/blob/23daf413794296cdb8fbe68beb16d965f3199517/google/cloud/bigtable/emulator/table.cc#L194)):*
> ```cpp
>           return maybe_cf.status();
>         }
>         cf = maybe_cf.value();
> ```

std::move(maybe_cf)



_Code quote:_
```cpp
maybe_cf
```
___
*[`google/cloud/bigtable/emulator/column_family.cc` line 172 at r11](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/3#-OQrnd659JSN9Z6Fx5MB:-OQrnd659JSN9Z6Fx5MC:b-7pk7gl) ([raw file](https://github.com/Unoperate/google-cloud-cpp/blob/23daf413794296cdb8fbe68beb16d965f3199517/google/cloud/bigtable/emulator/column_family.cc#L172)):*
> ```cpp
>     auto deleted_cells = column.second.DeleteTimeRange(time_range);
>     if (!deleted_cells.empty()) {
>       res[std::move(column.first)] = std::move(deleted_cells);
> ```

I believe this `std::move` doesn't change anything. If I'm right `column.first` is a `cref`, so after `std::move` it will become a `const&&` which will still not match to the move ctor.



_Code quote:_
```cpp
std::move(column.first)
```
___
*[`google/cloud/bigtable/emulator/column_family.cc` line 176 at r11](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/3#-OQrnSgA6v7snzc7g1ZT:-OQrnSgA6v7snzc7g1ZU:b-1xa04b) ([raw file](https://github.com/Unoperate/google-cloud-cpp/blob/23daf413794296cdb8fbe68beb16d965f3199517/google/cloud/bigtable/emulator/column_family.cc#L176)):*
> ```cpp
>   }
> 
>   rows_.erase(row_key);
> ```

`rows_.erase(row_it)` will save you additional lookup.



_Code quote:_
```cpp
rows_.erase(row_key);
```
___
*[`google/cloud/bigtable/emulator/column_family.h` line 239 at r11](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/3#-OQs1zB43Wvd0FBt1zk9:-OQs1zB43Wvd0FBt1zkA:b-wt2py9) ([raw file](https://github.com/Unoperate/google-cloud-cpp/blob/23daf413794296cdb8fbe68beb16d965f3199517/google/cloud/bigtable/emulator/column_family.h#L239)):*
> ```objc
>   // similar aggregate complex types. To construct an ordinary
>   // ColumnFamily, use the default constructor ColumnFamily().
>   friend StatusOr<std::shared_ptr<ColumnFamily>> ConstructAggregateColumnFamily(
> ```

Why not make it a static member?
___
*[`google/cloud/bigtable/emulator/column_family.h` line 280 at r11](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/3#-OQrgEWzErCiLEW8xtJD:-OQrgEWzErCiLEW8xtJE:bglnhxm) ([raw file](https://github.com/Unoperate/google-cloud-cpp/blob/23daf413794296cdb8fbe68beb16d965f3199517/google/cloud/bigtable/emulator/column_family.h#L280)):*
> ```objc
>       std::chrono::milliseconds timestamp, std::string const& value,
>       std::function<std::string(std::string const&, std::string const&)> const&
>           update_cell_fn);
> ```

The update function is an internal detail of `ColumnFamily`. IMO, we should not expose it to the user.

Instead, we should have two functions:
`SetCell()` - which would use the `DefaultUpdateFunction` and a `UpdateCell` which would use the one inferred from the schema. None of them would require the user to specify the function.



_Code quote:_
```objc
  absl::optional<std::string> SetCell(
      std::string const& row_key, std::string const& column_qualifier,
      std::chrono::milliseconds timestamp, std::string const& value,
      std::function<std::string(std::string const&, std::string const&)> const&
          update_cell_fn);
```
___
*[`google/cloud/bigtable/emulator/column_family.h` line 364 at r11](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/3#-OQrdh9t-lcw4tDXDNt4:-OQrdh9t-lcw4tDXDNt5:ba21f4l) ([raw file](https://github.com/Unoperate/google-cloud-cpp/blob/23daf413794296cdb8fbe68beb16d965f3199517/google/cloud/bigtable/emulator/column_family.h#L364)):*
> ```objc
> 
>   std::function<std::string(std::string const&, std::string const&)>
>   GetUpdateCell() {
> ```

According to https://google.github.io/styleguide/cppguide.html#Function_Names you can call it simply `update_cell()`. Also, I believe it should be marked `const`.

Also, shouldn't we get rid of this? This is an internal detail of this class. I believe if you follow my advice from the comment on `ColmnFamily::SetCell`, you will be able to remove it.



_Code quote:_
```objc
 GetUpdateCell
```
___
*[`google/cloud/bigtable/emulator/column_family.h` line 438 at r11](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/3#-OQraZga5oGT8KwFIviw:-OQraZga5oGT8KwFIvix:b-vmpj6u) ([raw file](https://github.com/Unoperate/google-cloud-cpp/blob/23daf413794296cdb8fbe68beb16d965f3199517/google/cloud/bigtable/emulator/column_family.h#L438)):*
> ```objc
> 
>   std::function<std::string(std::string const&, std::string const&)>
>       UpdateCell_ = DefaultUpdateCell;
> ```

Please follow the naming guideline. This should be `update_cell_`.



_Code quote:_
```objc
UpdateCell_
```


<!-- Sent from Reviewable.io -->

marete committed:
emulator: Make ConstructAggregateColumnFamily a static member instead.

Doh! Rather obvious in retrospect!

marete committed:
emulator: complex types: make it easier to update cells.

Updating a cell via the ColumnFamily should not require the user to
know about the update function. This commit fixes that.

marete committed:
emulator: mechanical formatting fixes.

marete committed:
emulator: make clang-tidy happy.

marete committed:
emulator: make clang-tidy happy.

marete committed:
emulator: clang-format.

marete committed:
emulator: remove unused getter.

marete committed:
emulator: fix for Google C++ style.

dopiera reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/3)* status: 0 of 11 files reviewed, 15 unresolved discussions (waiting on @marete)

___
*[`google/cloud/bigtable/emulator/column_family.cc` line 33 at r13](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/3#-ORDKOyL-WJyy68PlwVC:-ORDKOyL-WJyy68PlwVD:b88dw0q) ([raw file](https://github.com/Unoperate/google-cloud-cpp/blob/c789fc8d89bf2b103230980a3cd1bbd983bff601/google/cloud/bigtable/emulator/column_family.cc#L33)):*
> ```cpp
> absl::optional<std::string> ColumnRow::SetCell(
>     std::chrono::milliseconds timestamp, std::string const& value) {
>   absl::optional<std::string> ret = absl::nullopt;
> ```

This is incorrect (and was incorrect before) - it was my bug and I mislead you (sorry!). It was fine before - zero has no special meaning. Can you please revert this change so that we don't dig ourselves deeper into this bug?



_Code quote:_
```cpp
absl::optional<std::string> ret = absl::nullopt;
```


<!-- Sent from Reviewable.io -->

marete committed:
emulator: several changes for review.

marete reviewed (commented):
Thank you for your careful review!

PTAL.

> *[Reviewable](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/3#-:-ORDOMSi6rc_Yr2G-5GB:b6cuohw)* status: 0 of 11 files reviewed, 15 unresolved discussions (waiting on @dopiera)

___
*[`google/cloud/bigtable/emulator/column_family.h` line 239 at r11](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/3#-OQs1zB43Wvd0FBt1zk9:-ORCPUU-AstFxhCFh1ns:b-4xi60a) ([raw file](https://github.com/Unoperate/google-cloud-cpp/blob/23daf413794296cdb8fbe68beb16d965f3199517/google/cloud/bigtable/emulator/column_family.h#L239)):*
<details><summary><i>Previously, dopiera (Marek Dopiera) wrote…</i></summary><blockquote>

Why not make it a static member?
</blockquote></details>

Doh! Yes of course! It is mistakes like these that make me feel like a real C++ neophyte!
___
*[`google/cloud/bigtable/emulator/column_family.h` line 280 at r11](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/3#-OQrgEWzErCiLEW8xtJD:-ORDFKNLBHGGAHiLtoG0:b-896fix) ([raw file](https://github.com/Unoperate/google-cloud-cpp/blob/23daf413794296cdb8fbe68beb16d965f3199517/google/cloud/bigtable/emulator/column_family.h#L280)):*
<details><summary><i>Previously, dopiera (Marek Dopiera) wrote…</i></summary><blockquote>

The update function is an internal detail of `ColumnFamily`. IMO, we should not expose it to the user.

Instead, we should have two functions:
`SetCell()` - which would use the `DefaultUpdateFunction` and a `UpdateCell` which would use the one inferred from the schema. None of them would require the user to specify the function.
</blockquote></details>

Done.
___
*[`google/cloud/bigtable/emulator/column_family.h` line 364 at r11](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/3#-OQrdh9t-lcw4tDXDNt4:-ORDGUdJ3HVKtnAizCOc:b-896fix) ([raw file](https://github.com/Unoperate/google-cloud-cpp/blob/23daf413794296cdb8fbe68beb16d965f3199517/google/cloud/bigtable/emulator/column_family.h#L364)):*
<details><summary><i>Previously, dopiera (Marek Dopiera) wrote…</i></summary><blockquote>

According to https://google.github.io/styleguide/cppguide.html#Function_Names you can call it simply `update_cell()`. Also, I believe it should be marked `const`.

Also, shouldn't we get rid of this? This is an internal detail of this class. I believe if you follow my advice from the comment on `ColmnFamily::SetCell`, you will be able to remove it.
</blockquote></details>

Done.
___
*[`google/cloud/bigtable/emulator/column_family.h` line 438 at r11](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/3#-OQraZga5oGT8KwFIviw:-ORDH9U10v44sc3-mBYJ:b-896fix) ([raw file](https://github.com/Unoperate/google-cloud-cpp/blob/23daf413794296cdb8fbe68beb16d965f3199517/google/cloud/bigtable/emulator/column_family.h#L438)):*
<details><summary><i>Previously, dopiera (Marek Dopiera) wrote…</i></summary><blockquote>

Please follow the naming guideline. This should be `update_cell_`.
</blockquote></details>

Done.
___
*[`google/cloud/bigtable/emulator/column_family.cc` line 172 at r11](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/3#-OQrnd659JSN9Z6Fx5MB:-ORDJM_eCVntSUTlAtLx:b-896fix) ([raw file](https://github.com/Unoperate/google-cloud-cpp/blob/23daf413794296cdb8fbe68beb16d965f3199517/google/cloud/bigtable/emulator/column_family.cc#L172)):*
<details><summary><i>Previously, dopiera (Marek Dopiera) wrote…</i></summary><blockquote>

I believe this `std::move` doesn't change anything. If I'm right `column.first` is a `cref`, so after `std::move` it will become a `const&&` which will still not match to the move ctor.
</blockquote></details>

Done.
___
*[`google/cloud/bigtable/emulator/column_family.cc` line 176 at r11](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/3#-OQrnSgA6v7snzc7g1ZT:-ORDJvsT9Fb3Iw9IVn6z:b-896fix) ([raw file](https://github.com/Unoperate/google-cloud-cpp/blob/23daf413794296cdb8fbe68beb16d965f3199517/google/cloud/bigtable/emulator/column_family.cc#L176)):*
<details><summary><i>Previously, dopiera (Marek Dopiera) wrote…</i></summary><blockquote>

`rows_.erase(row_it)` will save you additional lookup.
</blockquote></details>

Done.
___
*[`google/cloud/bigtable/emulator/table.cc` line 194 at r11](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/3#-OQs3YGlBtcbx04fRmwl:-ORDLHXM9Vfa4_5LhC3j:bgp4h0k) ([raw file](https://github.com/Unoperate/google-cloud-cpp/blob/23daf413794296cdb8fbe68beb16d965f3199517/google/cloud/bigtable/emulator/table.cc#L194)):*
<details><summary><i>Previously, dopiera (Marek Dopiera) wrote…</i></summary><blockquote>

std::move(maybe_cf)
</blockquote></details>

Not sure I understand this one. Aren't we in fact dealing with a smart pointer? What's the point of avoiding a copy?


<!-- Sent from Reviewable.io -->

marete commented:
PTAL.

marete committed:
Merge branch 'fix_timestamp_handling' into topic-add_to_cell

marete committed:
Merge branch 'upstream_emulator' into topic-add_to_cell

marete reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/3)* status: 0 of 11 files reviewed, 15 unresolved discussions (waiting on @dopiera)

___
*[`google/cloud/bigtable/emulator/column_family.cc` line 33 at r13](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/3#-ORDKOyL-WJyy68PlwVC:-ORQyhG3EsbEOr9-4pdD:b-896fix) ([raw file](https://github.com/Unoperate/google-cloud-cpp/blob/c789fc8d89bf2b103230980a3cd1bbd983bff601/google/cloud/bigtable/emulator/column_family.cc#L33)):*
<details><summary><i>Previously, dopiera (Marek Dopiera) wrote…</i></summary><blockquote>

This is incorrect (and was incorrect before) - it was my bug and I mislead you (sorry!). It was fine before - zero has no special meaning. Can you please revert this change so that we don't dig ourselves deeper into this bug?
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

marete reviewed (commented):
All done (and I have merged in the 0 and negative timestamp handling fixes). PTAL.

> *[Reviewable](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/3#-:-ORQz2eA3Lhb_Xqi-2KR:b-2v8z3s)* status: 0 of 11 files reviewed, 15 unresolved discussions (waiting on @dopiera)



<!-- Sent from Reviewable.io -->

marete committed:
emulator: UpdateCell: Eliminate calls to std::abort().

If a value in a cell cannot be decoded as an int64, that is still no
reason to crash the entire server.

dopiera reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/3)* status: 0 of 11 files reviewed, 4 unresolved discussions (waiting on @marete)

___
*[`google/cloud/bigtable/emulator/table.cc` line 194 at r11](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/3#-OQs3YGlBtcbx04fRmwl:-ORvREhbBp0AuAmChmxE:bodn3tt) ([raw file](https://github.com/Unoperate/google-cloud-cpp/blob/23daf413794296cdb8fbe68beb16d965f3199517/google/cloud/bigtable/emulator/table.cc#L194)):*
<details><summary><i>Previously, marete (Brian Gitonga Marete) wrote…</i></summary><blockquote>

Not sure I understand this one. Aren't we in fact dealing with a smart pointer? What's the point of avoiding a copy?
</blockquote></details>

I did miss the fact that it was a `shared_ptr<>`, but copying `shared_ptr<>`s  is not free. Internally, it needs to first increase and then decrease (when the extra copy is destroyed) the internal counter. This counter is atomic and for atomicity to work, some portion of the CPU caches may need to be invalidated. It also prohibits the compiler and CPU to do certain reordering for optimization.
___
*[`google/cloud/bigtable/emulator/column_family.cc` line 58 at r16](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/3#-ORveC4g3NF-XtLTO8Js:-ORveC4g3NF-XtLTO8Jt:bpoai51) ([raw file](https://github.com/Unoperate/google-cloud-cpp/blob/8978ccf4c446f6844bed963785ac43753a643287/google/cloud/bigtable/emulator/column_family.cc#L58)):*
> ```cpp
>     }
>     ret = std::move(cell_it->second);
>     cells_[timestamp] = std::move(maybe_update_value.value());
> ```

This is an additional lookup. Why not `cell_it->second = std::move(maybe_update_value).value()`?

Also, you might as well `maybe_update_value.value().swap(cell_it->second);`



_Code quote:_
```cpp
cells_[timestamp]
```
___
*[`google/cloud/bigtable/emulator/table.cc` line 617 at r16](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/3#-ORvUd508VNvZL3YnZWR:-ORvUd508VNvZL3YnZWS:b-ajjjhk) ([raw file](https://github.com/Unoperate/google-cloud-cpp/blob/8978ccf4c446f6844bed963785ac43753a643287/google/cloud/bigtable/emulator/table.cc#L617)):*
> ```cpp
> Status RowTransaction::AddToCell(
>     ::google::bigtable::v2::Mutation_AddToCell const& add_to_cell,
>     absl::optional<std::chrono::milliseconds> timestamp_override) {
> ```

I think most of this function doesn't belong to the `RowTransaction` class. I think it needs to be moved to `Table` and only the last 15 lines should remain here.
___
*[`google/cloud/bigtable/emulator/column_family.h` line 357 at r16](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/3#-ORv_fL-5_brdcBsYCF7:-ORv_fL-5_brdcBsYCF8:bml9jkt) ([raw file](https://github.com/Unoperate/google-cloud-cpp/blob/8978ccf4c446f6844bed963785ac43753a643287/google/cloud/bigtable/emulator/column_family.h#L357)):*
> ```objc
> 
>   static StatusOr<std::string> DefaultUpdateCell(std::string const& /*existing_value*/,
>                                        std::string const& new_value) {
> ```

I think I'd change the signature to `(std::string&& existing_value, std::string&& new_value)` - this way you'll avoid copies here. Moreover, in `Max` and `Min` you can return the appropriate buffers directly and also avoid copies then.

Yet another approach to achieve the same goal is to do `(std::string& existing_value, std::string&& new_value)` and operate on existing value directly.



_Code quote:_
```objc
  static StatusOr<std::string> DefaultUpdateCell(std::string const& /*existing_value*/,
                                       std::string const& new_value) {
```


<!-- Sent from Reviewable.io -->

marete committed:
emulator: AddToCell: Fixes for review.

marete committed:
emulator: AddToCell: mechanical formatting fixes.

marete reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/3)* status: 0 of 11 files reviewed, 4 unresolved discussions (waiting on @dopiera)

___
*[`google/cloud/bigtable/emulator/column_family.h` line 357 at r16](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/3#-ORv_fL-5_brdcBsYCF7:-OSOgjr65uRZnXZxKNes:b-896fix) ([raw file](https://github.com/Unoperate/google-cloud-cpp/blob/8978ccf4c446f6844bed963785ac43753a643287/google/cloud/bigtable/emulator/column_family.h#L357)):*
<details><summary><i>Previously, dopiera (Marek Dopiera) wrote…</i></summary><blockquote>

I think I'd change the signature to `(std::string&& existing_value, std::string&& new_value)` - this way you'll avoid copies here. Moreover, in `Max` and `Min` you can return the appropriate buffers directly and also avoid copies then.

Yet another approach to achieve the same goal is to do `(std::string& existing_value, std::string&& new_value)` and operate on existing value directly.
</blockquote></details>

Done.
___
*[`google/cloud/bigtable/emulator/column_family.cc` line 58 at r16](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/3#-ORveC4g3NF-XtLTO8Js:-OSO_rl62DmblobosEaI:b-896fix) ([raw file](https://github.com/Unoperate/google-cloud-cpp/blob/8978ccf4c446f6844bed963785ac43753a643287/google/cloud/bigtable/emulator/column_family.cc#L58)):*
<details><summary><i>Previously, dopiera (Marek Dopiera) wrote…</i></summary><blockquote>

This is an additional lookup. Why not `cell_it->second = std::move(maybe_update_value).value()`?

Also, you might as well `maybe_update_value.value().swap(cell_it->second);`
</blockquote></details>

Done.
___
*[`google/cloud/bigtable/emulator/table.cc` line 194 at r11](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/3#-OQs3YGlBtcbx04fRmwl:-OSOaP1B49Ku0sl_wCnr:b-896fix) ([raw file](https://github.com/Unoperate/google-cloud-cpp/blob/23daf413794296cdb8fbe68beb16d965f3199517/google/cloud/bigtable/emulator/table.cc#L194)):*
<details><summary><i>Previously, dopiera (Marek Dopiera) wrote…</i></summary><blockquote>

I did miss the fact that it was a `shared_ptr<>`, but copying `shared_ptr<>`s  is not free. Internally, it needs to first increase and then decrease (when the extra copy is destroyed) the internal counter. This counter is atomic and for atomicity to work, some portion of the CPU caches may need to be invalidated. It also prohibits the compiler and CPU to do certain reordering for optimization.
</blockquote></details>

Done.
___
*[`google/cloud/bigtable/emulator/table.cc` line 617 at r16](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/3#-ORvUd508VNvZL3YnZWR:-OSOiTM4Cpfy0PUq2sYG:b69uxql) ([raw file](https://github.com/Unoperate/google-cloud-cpp/blob/8978ccf4c446f6844bed963785ac43753a643287/google/cloud/bigtable/emulator/table.cc#L617)):*
<details><summary><i>Previously, dopiera (Marek Dopiera) wrote…</i></summary><blockquote>

I think most of this function doesn't belong to the `RowTransaction` class. I think it needs to be moved to `Table` and only the last 15 lines should remain here.
</blockquote></details>

I plan to move all the logic except calls to SetCell or UpdateCell from RowTransacion for ALL the methods under TBL-63. Is that Okay?


<!-- Sent from Reviewable.io -->

marete reviewed (commented):
PTAL.

> *[Reviewable](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/3#-:-OSOnmwf0g2-ckEXyvFG:b-83330m)* status: 0 of 11 files reviewed, 4 unresolved discussions (waiting on @dopiera)



<!-- Sent from Reviewable.io -->

marete commented:
PTAL.

marete committed:
Merge branch 'upstream_emulator' into topic-add_to_cell

marete committed:
Merge branch 'upstream_emulator' into topic-add_to_cell

marete committed:
Fix typos.

dopiera reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/3)* status: 0 of 11 files reviewed, 4 unresolved discussions

___
*[`google/cloud/bigtable/emulator/column_family.h` line 357 at r16](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/3#-ORv_fL-5_brdcBsYCF7:-OT1M7u8AGCtaNCkwV7W:b-79asba) ([raw file](https://github.com/Unoperate/google-cloud-cpp/blob/8978ccf4c446f6844bed963785ac43753a643287/google/cloud/bigtable/emulator/column_family.h#L357)):*
<details><summary><i>Previously, marete (Brian Gitonga Marete) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

But you changed it to `std::string const&&` rather than `std::string &&`. This unfortunately will yield a copy. `const &&` is fishy in general - I don't think there are many situation where it makes sense.

Also, you didn't change the first parameter, preventing you from making the optimization in my next comment.
___
*[`google/cloud/bigtable/emulator/table.cc` line 617 at r16](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/3#-ORvUd508VNvZL3YnZWR:-OT1NbP0AK_AYB4WAJKA:b7e4hsb) ([raw file](https://github.com/Unoperate/google-cloud-cpp/blob/8978ccf4c446f6844bed963785ac43753a643287/google/cloud/bigtable/emulator/table.cc#L617)):*
<details><summary><i>Previously, marete (Brian Gitonga Marete) wrote…</i></summary><blockquote>

I plan to move all the logic except calls to SetCell or UpdateCell from RowTransacion for ALL the methods under TBL-63. Is that Okay?
</blockquote></details>

Makes sense, thanks
___
*[`google/cloud/bigtable/emulator/column_family.h` line 443 at r18](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/3#-OT1N5KU-AmUZfgOAeal:-OT1N5KV9G1c7I-6IGOG:b-jixsxh) ([raw file](https://github.com/Unoperate/google-cloud-cpp/blob/d6e5bdffcf7e8de749d839fe531636fe0d908a5e/google/cloud/bigtable/emulator/column_family.h#L443)):*
> ```objc
> 
>     if (existing_int.value() > new_int.value()) {
>       return google::cloud::internal::EncodeBigEndian(existing_int.value());
> ```

Why not return `existing_value` then?
___
*[`google/cloud/bigtable/emulator/column_family.h` line 446 at r18](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/3#-OT1NB7mCexeQGKjVEJ8:-OT1NB7mCexeQGKjVEJ9:bphp9cn) ([raw file](https://github.com/Unoperate/google-cloud-cpp/blob/d6e5bdffcf7e8de749d839fe531636fe0d908a5e/google/cloud/bigtable/emulator/column_family.h#L446)):*
> ```objc
>     }
> 
>     return google::cloud::internal::EncodeBigEndian(new_int.value());
> ```

Why not return `new_value`?
___
*[`google/cloud/bigtable/emulator/column_family.h` line 463 at r18](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/3#-OT1NGOI61g1eCyNvHkL:-OT1NGOI61g1eCyNvHkM:b-bn0wcg) ([raw file](https://github.com/Unoperate/google-cloud-cpp/blob/d6e5bdffcf7e8de749d839fe531636fe0d908a5e/google/cloud/bigtable/emulator/column_family.h#L463)):*
> ```objc
> 
>     if (existing_int.value() < new_int.value()) {
>       return google::cloud::internal::EncodeBigEndian(existing_int.value());
> ```

Why not `exising_value`?
___
*[`google/cloud/bigtable/emulator/column_family.h` line 466 at r18](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/3#-OT1NP1K2r-mwgI-IqPM:-OT1NP1K2r-mwgI-IqPN:b-9d1o05) ([raw file](https://github.com/Unoperate/google-cloud-cpp/blob/d6e5bdffcf7e8de749d839fe531636fe0d908a5e/google/cloud/bigtable/emulator/column_family.h#L466)):*
> ```objc
>     }
> 
>     return google::cloud::internal::EncodeBigEndian(new_int.value());
> ```

Why not `new_value`?


<!-- Sent from Reviewable.io -->

dopiera reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/3)* status: 0 of 11 files reviewed, 4 unresolved discussions (waiting on @marete)

___
*[`google/cloud/bigtable/emulator/column_family.h` line 357 at r16](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/3#-ORv_fL-5_brdcBsYCF7:-OT1hxXc2xrMM8ufpJLZ:b-d84in2) ([raw file](https://github.com/Unoperate/google-cloud-cpp/blob/8978ccf4c446f6844bed963785ac43753a643287/google/cloud/bigtable/emulator/column_family.h#L357)):*
<details><summary><i>Previously, dopiera (Marek Dopiera) wrote…</i></summary><blockquote>

But you changed it to `std::string const&&` rather than `std::string &&`. This unfortunately will yield a copy. `const &&` is fishy in general - I don't think there are many situation where it makes sense.

Also, you didn't change the first parameter, preventing you from making the optimization in my next comment.
</blockquote></details>

Unresolving the discussion.


<!-- Sent from Reviewable.io -->

marete committed:
Merge branch 'upstream_emulator' into topic-add_to_cell

marete committed:
emulator: mechanical formatting fixes by CI tool.

marete committed:
emulator: UpdateCell: eliminate buggy use of `std::string const&&'.

marete committed:
emulator: remove additional instances of buggy usage of `std::string const&&'.

marete committed:
emulator: AddToCell: Eliminate duplicate integer encodings in several places.

marete reviewed (commented):
All Done, PTAL.

> *[Reviewable](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/3#-:-OTWLrXx1zuicuSCtBzT:b-wcwjaz)* status: 0 of 12 files reviewed, 4 unresolved discussions (waiting on @dopiera)

___
*[`google/cloud/bigtable/emulator/column_family.h` line 357 at r16](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/3#-ORv_fL-5_brdcBsYCF7:-OTWFWRtE0zdBBB0KTwf:b-4acwvb) ([raw file](https://github.com/Unoperate/google-cloud-cpp/blob/8978ccf4c446f6844bed963785ac43753a643287/google/cloud/bigtable/emulator/column_family.h#L357)):*
<details><summary><i>Previously, dopiera (Marek Dopiera) wrote…</i></summary><blockquote>

Unresolving the discussion.
</blockquote></details>

The existing value is needed to return. We can either move it to the return value or move it to this function, not both. However, I have fixed the problem for the 2nd parameter and learned quite a bit about what `std::move()` actually means.
___
*[`google/cloud/bigtable/emulator/column_family.h` line 443 at r18](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/3#-OT1N5KU-AmUZfgOAeal:-OT3SmAM6i5CW9xyfH-o:b2c8gf) ([raw file](https://github.com/Unoperate/google-cloud-cpp/blob/d6e5bdffcf7e8de749d839fe531636fe0d908a5e/google/cloud/bigtable/emulator/column_family.h#L443)):*
<details><summary><i>Previously, dopiera (Marek Dopiera) wrote…</i></summary><blockquote>

Why not return `existing_value` then?
</blockquote></details>

This is a good catch. Thanks. Done.
___
*[`google/cloud/bigtable/emulator/column_family.h` line 446 at r18](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/3#-OT1NB7mCexeQGKjVEJ8:-OT3SpMD1Yqmhd9s5166:b-896fix) ([raw file](https://github.com/Unoperate/google-cloud-cpp/blob/d6e5bdffcf7e8de749d839fe531636fe0d908a5e/google/cloud/bigtable/emulator/column_family.h#L446)):*
<details><summary><i>Previously, dopiera (Marek Dopiera) wrote…</i></summary><blockquote>

Why not return `new_value`?
</blockquote></details>

Done.
___
*[`google/cloud/bigtable/emulator/column_family.h` line 463 at r18](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/3#-OT1NGOI61g1eCyNvHkL:-OT3VJGJ3fX0k33Uyuzl:b-9pbmlp) ([raw file](https://github.com/Unoperate/google-cloud-cpp/blob/d6e5bdffcf7e8de749d839fe531636fe0d908a5e/google/cloud/bigtable/emulator/column_family.h#L463)):*
<details><summary><i>Previously, dopiera (Marek Dopiera) wrote…</i></summary><blockquote>

Why not `exising_value`?
</blockquote></details>

Thank you for this good catch. Fixed.
___
*[`google/cloud/bigtable/emulator/column_family.h` line 466 at r18](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/3#-OT1NP1K2r-mwgI-IqPM:-OT3VOkn2wAXiA7_ik6E:b-896fix) ([raw file](https://github.com/Unoperate/google-cloud-cpp/blob/d6e5bdffcf7e8de749d839fe531636fe0d908a5e/google/cloud/bigtable/emulator/column_family.h#L466)):*
<details><summary><i>Previously, dopiera (Marek Dopiera) wrote…</i></summary><blockquote>

Why not `new_value`?
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

marete commented:
PTAL.

dopiera reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/3)* status: 0 of 12 files reviewed, all discussions resolved

___
*[`google/cloud/bigtable/emulator/column_family.h` line 357 at r16](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/3#-ORv_fL-5_brdcBsYCF7:-OTXAbLuAXAEgMpUgcxO:bilfp83) ([raw file](https://github.com/Unoperate/google-cloud-cpp/blob/8978ccf4c446f6844bed963785ac43753a643287/google/cloud/bigtable/emulator/column_family.h#L357)):*
<details><summary><i>Previously, marete (Brian Gitonga Marete) wrote…</i></summary><blockquote>

The existing value is needed to return. We can either move it to the return value or move it to this function, not both. However, I have fixed the problem for the 2nd parameter and learned quite a bit about what `std::move()` actually means.
</blockquote></details>

Ah, that makes sense, thank you.


<!-- Sent from Reviewable.io -->

dopiera reviewed (approved):
> Reviewed 2 of 10 files at r7, 2 of 6 files at r8, 1 of 6 files at r9, 1 of 5 files at r10, 6 of 6 files at r19, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/Unoperate/google-cloud-cpp/3)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @marete)



<!-- Sent from Reviewable.io -->
